### PR TITLE
Fix: Improve missing editor scenario

### LINKED ISSUE #1689: Open editor improvements
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

* feedback when the editor fails to open for some reason
* Deal with `$EDITOR` not set and `code` not being installed on the `$PATH`. Guide towards installing vscode, and link to https://code.visualstudio.com/docs/setup/mac#_launching-from-the-command-line

We can benchmark and take inspiration from:

* `launchEditor` in [react-dev-utils](https://www.npmjs.com/package/react-dev-utils) seems to handle cases like this better
* playwright allows opening vscode in its ui mode

  Image

* next.js redbox callstack items link to code editor
   Image

* Add documentation + link to it when opening a code editor

### Examples 🌈

_No response_

### Motivation 🔦

_No response_

### PR DESCRIPTION
**Pull request by bharatkashyap**
-  Closes #1689 

- A missing `code` inside `PATH` shows the following error: 
Screenshot 2023-06-28 at 2 50 47 PM

which links to the following docs page:

Screenshot 2023-06-28 at 2 51 45 PM

- Do we add additional sections for other editors, and a catch all default (restart Toolpad and try again)? Scouting for suggestions here.

- [x] I've read and followed the [contributing guide](https://github.com/mui/mui-toolpad/blob/master/CONTRIBUTING.md#sending-a-pull-request) on how to create great pull requests.
- [x] I've updated the relevant documentation for any new or updated feature
- [x] I've linked relevant GitHub issue with "Closes #<issue id>"
- [x] I've added a visual demonstration under the form of a screenshot or video

### CODE REVIEW

File: packages/toolpad-app/src/toolpadDataSources/local/client.tsx [hunk 1]
```diff
@@ -83,12 +95,16 @@ function QueryEditor({
...
+    setMissingEditor(false);
+  }, []);
+
   const openEditor = React.useCallback(() => {
     fetchPrivate({ kind: 'openEditor' }).catch((err) => {
-      // TODO: Write docs with instructions on how to install editor
-      // Add a good looking alert box and inline some instructions and link to docs
-      // eslint-disable-next-line no-alert
-      alert(err.message);
+      console.error(err);
```
Comment (Janpot) [inline packages/toolpad-app/src/toolpadDataSources/local/client.tsx:106]:
```suggestion
```

File: packages/toolpad-app/src/toolpadDataSources/local/client.tsx [hunk 2]
```diff
@@ -113,74 +129,99 @@ function QueryEditor({
...
-      </QueryPreview>
-    </SplitPane>
+        </QueryInputPanel>
+
+        <QueryPreview isLoading={previewIsLoading} error={preview?.error}>
+          <JsonView sx={{ height: '100%' }} copyToClipboard src={preview?.data} />
+        </QueryPreview>
+      </SplitPane>
+
+      <Dialog
```
Comment (Janpot) [inline packages/toolpad-app/src/toolpadDataSources/local/client.tsx:202]: You'll have to extract and reuse the logic for https://github.com/mui/mui-toolpad/blob/b083a524b91dc35eb2572412b05aa8abb4930a97/packages/toolpad-app/src/toolpad/AppEditor/HierarchyExplorer/CreateCodeComponentNodeDialog.tsx#L117C1-L117C1

Comment (bharatkashyap) [reply packages/toolpad-app/src/toolpadDataSources/local/client.tsx:202]: Agreed, we should have a reusable Button which allows opening the editor. I'll refactor this to allow reuse and some more configurability

File: packages/toolpad-app/src/components/OpenCodeEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+interface OpenCodeEditorButtonProps extends ButtonProps {
+  filePath: string;
+  fileType: string;
+  iconButton?: boolean;
+}
+
+interface MissingEditorDialogProps {
+  open: boolean;
+  setMissingEditor: React.Dispatch<React.SetStateAction<boolean>>;
```
Comment (Janpot) [inline packages/toolpad-app/src/components/OpenCodeEditor.tsx:25]: Let's use the convention of `open` and `onClose` for dialogs?

File: packages/toolpad-app/src/components/OpenCodeEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  }, [setMissingEditor]);
+
+  return (
+    <Dialog
+      open={open}
+      onClose={handleMissingEditorDialogClose}
+      aria-labelledby="alert-dialog-title"
+      aria-describedby="alert-dialog-description"
+    >
+      <DialogTitle id="alert-dialog-title">{'Editor not found'}</DialogTitle>
```
Comment (Janpot) [inline packages/toolpad-app/src/components/OpenCodeEditor.tsx:40]: I usually use 

```tsx
const id = React.id()
// ...
<DialogTitle id={`${id}-title`}>
``
to guarantee uniqueness

File: packages/toolpad-app/src/components/OpenCodeEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      setMissingEditorDialog(true);
+    });
+  }, [filePath, fileType]);
+
+  return (
+    <React.Fragment>
+      {iconButton ? (
+        <Tooltip title="Open in code editor">
+          <IconButton size="small" onClick={handleClick} {...rest}>
+            <CodeIcon fontSize="small" color="primary" />
```
Comment (Janpot) [inline packages/toolpad-app/src/components/OpenCodeEditor.tsx:78]: I recently learnt that you can do
```suggestion
            <CodeIcon fontSize="inherit" color="primary" />
```
to have it follow the `IconButton` size

File: packages/toolpad-app/src/server/localMode.ts [hunk 6]
```diff
@@ -816,8 +817,8 @@ export async function openCodeEditor(filePath: string, fileType: string) {
   let resolvedPath = filePath;
 
   if (fileType === 'query') {
-    const resourcesFolder = getResourcesFolder(root);
-    resolvedPath = getFunctionFilePath(resourcesFolder, filePath);
+    const resourcesFolder = await getResourcesFolder(root);
+    resolvedPath = await getFunctionFilePath(resourcesFolder, filePath);
```
Comment (Janpot) [inline packages/toolpad-app/src/server/localMode.ts:821]: If you move this whole function into the class you can do

```tsx
resolvedPath = this.functionsManager.getFunctionFilePath(filePath)
```

Comment (bharatkashyap) [reply packages/toolpad-app/src/server/localMode.ts]: I need to also be able to expose this function to be able to import it in `rpc.ts`, is there a way to do that while also moving this function inside the class?

Comment (Janpot) [reply packages/toolpad-app/src/server/localMode.ts]: See https://github.com/mui/mui-toolpad/blob/master/packages/toolpad-app/src/server/liveProject.ts

**Review by Janpot — APPROVED** [review]

**Pull request merged**